### PR TITLE
Async IO Support for XDP

### PR DESCRIPTION
**Pull request by nibanks**
## Description

- Updates the XDP datapath layer to support async IO (pending a RX/TX)
- Adds performance improvements to the polling loop logic

## Testing

Automation - BVT & Perf

## Documentation

N/A

### CODE REVIEW

File: src/platform/unittest/DataPathTest.cpp [hunk 1]
```diff
@@ -731,6 +756,48 @@ TEST_P(DataPathTest, UdpData)
...
     memcpy(ClientBuffer->Buffer, ExpectedData, ExpectedDataSize);
 
+    VERIFY_QUIC_SUCCESS(Client.Send(ClientSendData));
+    auto success = CxPlatEventWaitWithTimeout(RecvContext.ClientCompletion, 2000);
+#ifndef WIN32
+    if (!success && GetParam() == 4) { // Generate live dump
+        system("pwsh -Command 'Get-StorageDiagnosticInfo -StorageSubSystemFriendlyName (Get-StorageSubSystem)[0].FriendlyName -IncludeLiveDump -DestinationPath D:\\a\\1\\s\\artifacts\\logs\\msquicplatformtest.exe_Xdp'")
+    }
+#endif
+    ASSERT_TRUE(success);
```
Comment (nibanks) [inline src/platform/unittest/DataPathTest.cpp:766]: TODO: Remove once we have a dump.

File: src/manifest/MsQuic.wprp [hunk 2]
```diff
@@ -176,6 +176,7 @@
         <EventCollectorId Value="EC_LowVolume">
           <EventProviders>
             <EventProviderId Value="EP_MsQuicEtw_LowVolume" />
+            <EventProviderId Value="EP_XdpEtwPerFrame" />
```
Comment (nibanks) [inline src/manifest/MsQuic.wprp:179]: Remove these additions before merging.

File: src/platform/platform_worker.c [hunk 3]
```diff
@@ -184,6 +184,10 @@ CxPlatWorkersLazyStart(
             goto Error;
         }
         CxPlatWorkers[i].InitializedUpdatePollSqe = TRUE;
+#endif
+#ifdef WIN32
+        CxPlatWorkers[i].EventQ.WakeEvent = &CxPlatWorkers[i].WakeSqe;
+        CxPlatWorkers[i].EventQ.WakeEventUserData = (void*)&WorkerWakeEventPayload;
 #endif
```
Comment (nibanks) [inline src/platform/platform_worker.c:191]: Make all these changes cross-platform?

Comment (nibanks) [reply src/platform/platform_worker.c:191]: I'll clean this up in a follow up PR.

File: src/platform/datapath_raw_xdp.c [hunk 4]
```diff
@@ -1588,14 +1640,14 @@ CxPlatXdpTx(
         }
     }
 
-    if (XskRingError(&Queue->TxRing) && !Queue->Error) {
+    /*if (XskRingError(&Queue->TxRing) && !Queue->Error) {
```
[the remark this answers is not in the document]
Comment (nibanks) [reply src/platform/datapath_raw_xdp.c:1643]: 1. It was using a printf
2. It has never been hit in practice

So I just commented it out for now.

File: src/inc/quic_platform_winuser.h [hunk 5]
```diff
@@ -605,7 +621,39 @@ CxPlatEventQCleanup(
...
+}
+
+inline
+void
+CxPlatEventQSetRunning(
+    _In_ CXPLAT_EVENTQ* queue,
+    _In_ uint32_t OwningThread
+    )
+{
+    queue->OwningThread = OwningThread;
```
[the remark this answers is not in the document]
Comment (nibanks) [reply src/inc/quic_platform_winuser.h:634]: I was thinking I would need it for future. For instance, in TX path, if on the OwningThread, then we don't need to lock.

[the remark this answers is not in the document]
Comment (mtfriesen) [reply src/inc/quic_platform_winuser.h:634]: Sure, makes sense.

**Review by mtfriesen — APPROVED** [review]

**Review by csujedihy — APPROVED** [review]
:shipit:

**Review by mtfriesen — APPROVED** [review]

**Pull request merged**